src/utils/registry.ts

```diff
@@ -232,123 +207,176 @@ function readDescriptors(packagePath: string) {
-    const urlTemplate = DOWNLOAD_URLS[browserName][hostPlatform];
-    assert(urlTemplate, `ERROR: Playwright does not support ${browserName} on ${hostPlatform}`);
-    return util.format(urlTemplate, downloadHost, browser.revision);
+    const webkit = descriptors.find(d => d.name === 'webkit')!;
+    const webkitExecutable = executablePath(webkit.dir, 'webkit');
+    const webkitLinuxLddDirectories = [
+      path.join('minibrowser-gtk'),
+      path.join('minibrowser-gtk', 'bin'),
+      path.join('minibrowser-gtk', 'lib'),
+      path.join('minibrowser-wpe'),
+      path.join('minibrowser-wpe', 'bin'),
+      path.join('minibrowser-wpe', 'lib'),
+    ];
+    this._executables.push({
+      type: 'browser',
+      name: 'webkit',
+      browserName: 'webkit',
+      directoryIfExists: () => directoryIfExists(webkit.dir),
+      executablePathIfExists: () => executablePathIfExists(webkitExecutable),
+      installType: webkit.installByDefault ? 'download-by-default' : 'download-on-demand',
+      validateHostRequirements: () => this._validateHostRequirements('webkit', webkit.dir, webkitLinuxLddDirectories, ['libGLESv2.so.2', 'libx264.so'], ['']),
+      _download: () => this._downloadExecutable(webkit, webkitExecutable, DOWNLOAD_URLS['webkit'][hostPlatform], 'PLAYWRIGHT_WEBKIT_DOWNLOAD_HOST'),
+      _requires: [],
+    });
+
+    const ffmpeg = descriptors.find(d => d.name === 'ffmpeg')!;
+    const ffmpegExecutable = executablePath(ffmpeg.dir, 'ffmpeg');
+    this._executables.push({
+      type: 'tool',
+      name: 'ffmpeg',
+      browserName: undefined,
+      directoryIfExists: () => directoryIfExists(ffmpeg.dir),
+      executablePathIfExists: () => executablePathIfExists(ffmpegExecutable),
+      installType: ffmpeg.installByDefault ? 'download-by-default' : 'download-on-demand',
+      validateHostRequirements: () => Promise.resolve(),
+      _download: () => this._downloadExecutable(ffmpeg, ffmpegExecutable, DOWNLOAD_URLS['ffmpeg'][hostPlatform], 'PLAYWRIGHT_FFMPEG_DOWNLOAD_HOST'),
+      _requires: [],
+    });
   }
 
-  isSupportedBrowser(browserName: string): boolean {
-    // We retain browsers if they are found in the descriptor.
-    // Note, however, that there are older versions out in the wild that rely on
-    // the "download" field in the browser descriptor and use its value
-    // to retain and download browsers.
-    // As of v1.10, we decided to abandon "download" field.
-    return this._descriptors.some(browser => browser.name === browserName);
+  findExecutable(name: BrowserName): Executable;
+  findExecutable(name: string): Executable | undefined;
+  findExecutable(name: string): Executable | undefined {
+    return this._executables.find(b => b.name === name);
   }
 
-  private _installByDefault(): BrowserName[] {
-    return this._descriptors.filter(browser => browser.installByDefault).map(browser => browser.name);
+  private _addRequirementsAndDedupe(executables: Executable[] | undefined): ExecutableImpl[] {
```

Comment:
nit: make function instead of a method? I'd pop the no-executables usecase to the call sites. It'll make the `install` and `installDeps` API more clear since the default value will be readily available.

Reply:
Yep, doing that in a follow up.